Web Inspector: Network tab drop zone activates incorrectly

mirnovov opened the pull request:
#### 2c09c005dd59294f1251a0007ab1e9522717235c
<pre>
Web Inspector: Network tab drop zone activates incorrectly
<a href="https://bugs.webkit.org/show_bug.cgi?id=260251">https://bugs.webkit.org/show_bug.cgi?id=260251</a>

Reviewed by Devin Rousso.

This PR fixes a drop zone related issues caused by ImageResourceContentView being in NetworkTabContentView.
When dragging an item in when the image preview is active, both the HAR and Local Override drop zones activate, so modify DropZoneView so only one drop zone ever activates at a time.

Combined changes:
* Source/WebInspectorUI/UserInterface/Views/ImageResourceContentView.js:
(WI.ImageResourceContentView.prototype.contentAvailable):
* Source/WebInspectorUI/UserInterface/Views/NetworkResourceDetailView.js:
(WI.NetworkResourceDetailView.prototype.showContentViewForIdentifier):
(WI.NetworkResourceDetailView):
* Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
(WI.ResourceClusterContentView.prototype._contentViewForResourceType):

Canonical link: <a href="https://commits.webkit.org/270108@main">https://commits.webkit.org/270108@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b2ab3f3979217c6d923ad64217aceda52991619

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17821 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18813 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15933 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20586 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17493 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18154 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19628 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15449 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22161 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15800 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15616 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19939 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16218 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13749 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15376 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5893 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -180,6 +180,7 @@ WI.loaded = function()
     WI.modifierKeys = {altKey: false, metaKey: false, shiftKey: false};
     WI.visible = false;
     WI._windowKeydownListeners = [];
+    WI.dropActive = false;
```
rather than declaring this globally, how about having this be `WI.DropZoneView._active` so it's a bit more contained (and since it's only used by `WI.DropZoneView`)?

another option would be to do something similar to how `WI.ImageResourceContentView` has `disableInteractions` and propagate that when viewing the Network Tab.  tho this might be overly preventative (i.e. we may want to allow developers to drag over the `WI.ImageResourceContentView` to create a response local override)

i suggest the above because i think the only way that the `WI.DropZoneView` inside `WI.ImageResourceContentView` can get activated is if the user somehow starts their drag inside the `WI.ImageResourceContentView` instead of the `WI.NetworkTabContentView` (e.g. have a Finder window perfectly positioned to do so), and i dont think we want that kind of inconsistency

mirnovov replied:
> i suggest the above because i think the only way that the WI.DropZoneView inside WI.ImageResourceContentView can get activated is if the user somehow starts their drag inside the WI.ImageResourceContentView instead of the WI.NetworkTabContentView (e.g. have a Finder window perfectly positioned to do so), and i dont think we want that kind of inconsistency

Good point, done.

>(i.e. we may want to allow developers to drag over the WI.ImageResourceContentView to create a response local override)

Personally I think that since the user can still create local overrides via the 📑button its not a huge deal.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DropZoneView.js:
```diff
@@ -112,10 +112,11 @@ WI.DropZoneView = class DropZoneView extends WI.View
             return;
 
         this._startActiveDrag();
+        WI.dropActive = true;
```
can we move this into `_startActiveDrag`?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/DropZoneView.js:
```diff
@@ -127,6 +128,7 @@ WI.DropZoneView = class DropZoneView extends WI.View
             return;
 
         this._stopActiveDrag();
+        WI.dropActive = false;
```
can we move this into `_stopActiveDrag`?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NetworkTabContentView.js:
```diff
@@ -120,7 +120,7 @@ WI.NetworkTabContentView = class NetworkTabContentView extends WI.TabContentView
     dropZoneShouldAppearForDragEvent(dropZone, event)
     {
-        return event.dataTransfer.types.includes("Files");
+        return event.dataTransfer.types.includes("Files") && !event.dataTransfer.types.includes("text/uri-list");
```
i think we should split this into a separate PR as im not entirely sure we want to disallow this (i.e. what if the URI points to JSON?)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ImageResourceContentView.js:
```diff
@@ -25,7 +25,7 @@
 WI.ImageResourceContentView = class ImageResourceContentView extends WI.ResourceContentView
 {
-    constructor(resource, {disableInteractions} = {})
+    constructor(resource, {disableInteractions, disableDropping} = {})
```
nit: i think `disableDropZone` would be more apt

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/NetworkResourceDetailView.js:
```diff
@@ -121,7 +121,9 @@ WI.NetworkResourceDetailView = class NetworkResourceDetailView extends WI.Networ
             if (!this._resourceContentView)
-                this._resourceContentView = this._contentBrowser.showContentViewForRepresentedObject(this.representedObject);
+                this._resourceContentView = this._contentBrowser.showContentViewForRepresentedObject(
+                    this.representedObject, false, {inNetworkView: true}
```
style: please create a `const cookie = null;` variable instead of inlining `false`
style: please put all of this back on one line
```
if (!this._resourceContentView) {
    const cookie = null;
    this._resourceContentView = this._contentBrowser.showContentViewForRepresentedObject(this.representedObject, cookie, {disableDropZone: true});
}
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
```diff
@@ -25,13 +25,14 @@
 WI.ResourceClusterContentView = class ResourceClusterContentView extends WI.ClusterContentView
 {
-    constructor(resource)
+    constructor(resource, {inNetworkView} = {})
```
i think we should match the parameter name used by `WI.ImageResourceContentView` (i.e. `disableDropZone`) to allow for future engineers searching for `disableDropZone` to easily see the entire codepath in the search results instead of having to do a second search

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
```diff
@@ -263,7 +264,10 @@ WI.ResourceClusterContentView = class ResourceClusterContentView extends WI.Clus
+            if (this._inNetworkView)
+                return new WI.ImageResourceContentView(this._resource, {disableDropping: true});
+            else
+                return new WI.ImageResourceContentView(this._resource);
```
you can inline all of this to make it easier
```
return new WI.ImageResourceContentView(this._resource, {disableDropZone: this._disableDropZone});
```

mirnovov committed:
Web Inspector: Network tab drop zone activates incorrectly
https://bugs.webkit.org/show_bug.cgi?id=260251

Reviewed by Devin Rousso.

This PR fixes a drop zone related issues caused by ImageResourceContentView being in NetworkTabContentView.
When dragging an item in when the image preview is active, both the HAR and Local Override drop zones activate, so modify DropZoneView so only one drop zone ever activates at a time.

Combined changes:
* Source/WebInspectorUI/UserInterface/Views/ImageResourceContentView.js:
(WI.ImageResourceContentView.prototype.contentAvailable):
* Source/WebInspectorUI/UserInterface/Views/NetworkResourceDetailView.js:
(WI.NetworkResourceDetailView.prototype.showContentViewForIdentifier):
(WI.NetworkResourceDetailView):
* Source/WebInspectorUI/UserInterface/Views/ResourceClusterContentView.js:
(WI.ResourceClusterContentView.prototype._contentViewForResourceType):

Canonical link: https://commits.webkit.org/270108@main
